File: Leetcode 3701-3800/3755/3755.py
# Leetcode 3755: Find Maximum Balanced XOR Subarray Length
# https://leetcode.com/problems/find-maximum-balanced-xor-subarray-length/
# Solved on 26th of December, 2025
class Solution:
    def maxBalancedSubarray(self, nums: list[int]) -> int:
        """
        Finds the maximum length of a balanced XOR subarray.

        A subarray is balanced if the count of even numbers equals the count of odd numbers.

        :param nums: A list of integers.
        :return: The maximum length of a balanced XOR subarray.
        """
        seenStates = {(0, 0): -1}

        currentXor = 0
        balance = 0
        maxLength = 0
        for i, num in enumerate(nums):
            currentXor ^= num
            if num % 2 == 0:
                balance += 1
            else:
                balance -= 1

            currentState = (currentXor, balance)
            if currentState in seenStates:
                currentLength = i - seenStates[currentState]
                if currentLength > maxLength:
                    maxLength = currentLength
            else:
                seenStates[currentState] = i

        return maxLength

File: Leetcode 3701-3800/3755/3755.java
// Leetcode 3755: Find Maximum Balanced XOR Subarray Length
// https://leetcode.com/problems/find-maximum-balanced-xor-subarray-length/
// Solved on 26th of December, 2025
import java.util.HashMap;
import java.util.Map;

class Solution {
    /**
     * Finds the maximum length of a "balanced XOR subarray".
     * A subarray is balanced if the count of odd numbers equals the count of even numbers,
     * and the XOR sum of its elements is 0.
     * @param nums The input array of integers.
     * @return The maximum length of a balanced XOR subarray.
     */
    public int maxBalancedSubarray(int[] nums) {
        Map<Long, Integer> map = new HashMap<>();
        int n = nums.length;
        int xor = 0;
        int diff = 0;
        int maxLen = 0;
        int offset = n + 1;

        long initialKey = ((long) 0 << 32) | (0 + offset);
        map.put(initialKey, -1);

        for (int i = 0; i < n; i++) {
            xor ^= nums[i];
            
            if (nums[i] % 2 != 0) {
                diff++;
            } else {
                diff--;
            }

            long key = ((long) xor << 32) | (diff + offset);

            if (map.containsKey(key)) {
                maxLen = Math.max(maxLen, i - map.get(key));
            } else {
                map.put(key, i);
            }
        }
        return maxLen;
    }
}